### ISSUE TITLE
Conditional generation for an instrument

### ISSUE THREAD

**Issue by m41w4r3exe** [jthickstun/anticipation#9]
Thank you for sharing your work! Nice concept of 'anticipation'.

I wanted to experiment with generating notes, by choosing a specific instruments to generate notes for. However I couldn't find a way to do that. Is there a way to condition on a specfic instrument at any generation time?

For instance having an input of piano and drums, is it possible to generate specifically bass next to it that plays at the same time or after them?

**Comment by jthickstun** [comment]
This has been by far the most requested feature for these models. I hope to be able to provide this as a built-in feature of the next generation of models.

In the meantime, I can give some pointers to hack this functionality, although it will require a bit of work and isn't perfect. The idea is to mask the model logits for instruments that you don't want to generate. I've written some code on an experimental branch here:

https://github.com/jthickstun/anticipation/blob/84c4374277a7a6f598ff5a149a4a52547790feb5/anticipation/sample.py#L118

Notice the addition of the `masked_instrs=[]` input. You can pass in a restrictive list like below to, e.g., to create a mask that prevents generation of any instrument except violin (MIDI code 40):

```
masked_instrs = list(range(128))
masked_instrs.pop(40)
```

Note that the experimental branch here is quite messy, so I would recommend back-porting this specific logic to main.

**Comment by m41w4r3exe** [comment]
Thanks for fast reply and explanation! 

I was thinking of such a hack as well, but I was not sure whether this would create an undesired bias that would affect output quality. I feel like, it might unfortunately affect the quality of results, unless augmenting on training data has been done with a masking logic like this for instruments as well, I guess

I will try it out

**Comment by m41w4r3exe** [comment]
@jthickstun 

I tried to implement this way of generating, however I couldn't make it work.

The problem is that generation is stuck on a loop, where the same compound keeps on being created (same time, duration, instrument, pitch) without moving forward in time.

Any ideas how to go around it? Could you generate notes consistently via this hack?

**Comment by jthickstun** [comment]
Yes I've gotten this to work. I will try to find some time this upcoming week to backport the masking logic to main and share a working colab notebook.

In the meantime, here's the jupyter notebook I was using to do this: adding a violin part to Fur Elise (I included the midi copy of Fur Elise too to help with reproducibility).

[share.zip](https://github.com/jthickstun/anticipation/files/13624947/share.zip)

**Comment by m41w4r3exe** [comment]
thanks for sharing the notebook.

is that a different model you are using there as "/futile-think-tank-272/" ?

Because, from all my trials, there is always lots of same compounds repeated (same time, duration, instrument, pitch) and every 3-4 generation it is stuck on the same compound for over 20 generation step in a row before moving on to next time block.

Also, there are plenty of compounds generated with 0 duration.

Here is a notebook I was using to reproduce, running with a fork from your repo and only added the logic you have explained above and an example midi that is not included in LAKH dataset.

https://colab.research.google.com/drive/1j084bsbtA4F3dk6chvE_UeuIIWKTWh5u?usp=sharing

Also when logits are masked to condition for an instrument, the musicality of generated notes became really low; although it is very impressive when not conditioned on any instrument, definitely one of the best symbolic music generation I tried so far!

**Comment by jthickstun** [comment]
> is that a different model you are using there as "/futile-think-tank-272/" ?

That was our internal code name (weights & biased ID) for the publicly released medium-800k model.

Sorry to hear that my hack doesn't seem to be working for you; agreed this is definitely a bit out of scope for the current models. As I mentioned, built-in support for instrument conditioning is on my roadmap; I will be sure to let you know when a new model is available!
